### PR TITLE
set persistent, friendly hostname

### LINKED ISSUE #1231: configure hostnames on mantl nodes
We've seen issues in some environments where the hostnames of mantl nodes are not resolvable within a cluster. Some apps call the equiv of the gethostname api and, on some clouds, get something back that can’t be resolved [1]. Or, get a hostname that is unfamiliar (something like ip-10-1-1-57 on aws, for example) showing up in logs, UIs, etc.

Should we set and enforce a hostname based on the Mantl node name? This would have to persist across reboots.

[1] https://github.com/CiscoCloud/mantl/issues/1195

### PR DESCRIPTION
**Pull request by ryane**
- [x] Installs cleanly on a fresh build of most recent master branch
- [x] Upgrades cleanly from the most recent release
- [x] Updates documentation relevant to the changes

fixes #1231

---

* tested on AWS only so far. 
* will this work on other clouds?
* is this the best way to do it?

### CODE REVIEW

**Comment by kbroughton** [comment]
inventory_hostname is not particularly well defined 
Depending on the cloud provider and the settings of the ec2.ini or equiv you get different inventory_hostname's.  I think some people use terraform to create, and ansible ec2.py inventory to run mant.yml + their customization which may add some constraints.  I presume what you want is the prompt to show a friendly hostname like
x-worker-001> 

It might be best to set a custom fact that will not change the way inventory_hostname might.
I think it would be nice to also have the fqdn and /etc/hosts set correctly
cat /etc/hosts

10.x.y.z  <short-name>-worker-001  <short-name>-worker-001.<domain>

This would give more options for solving issuing certs and zookeeper connection strings etc.

While we're at it, why are the dns records generated of the form <short-name>-worker-001.node.<domain>?  Why is the .node. inserted?
It would be nice if we could have the same name at the prompt, in /etc/hosts and in the dns records by default, not having to fix them up post mant.yml

**Comment by ryane** [comment]
The shell prompt would be nice but that is only a part of it. See #1231, #1367, #1195 

What is the use case for `ec2.py`? I believe you get a reliable inventory_hostname when using `terraform.py`.

What source would we use to set the fact if we don't use inventory_hostname?

thanks for the feedback!

**Comment by kbroughton** [comment]
possibly ansible_fqdn or ansible_hostname.

A use case I've had was my first pass at mantl.
I needed to run some of my roles against my mantl cloud.  Terraform.py gives only a small subset of vars that ec2.py returns https://github.com/lorin/ansible-quickref/blob/master/ec2.rst.  Here is another person trying to do something similar
https://github.com/CiscoCloud/mantl/issues/1274#event-590692969

mantl deploys without a bastion or openvpn at the moment.  Thus you need to set 
destination_variable = private_dns_name
in ec2.ini

another reason, ec2.py is intended to be hacked to suit your needs.  So we have some custom code that parses aws tags so that we get groups of tag values, not role=blah, and that parses a comma separated list allowing us to tag in aws
role: tag1,tag2,tag3
and get hosts put into all those groups, so now are roles look normal, not like "role=control".
Not that that's a common thing, but there's lots of reasons an ansible power user might want to run ec2.py rather that terraform.py.

File: roles/common/tasks/main.yml [hunk 1]
```diff
@@ -18,6 +18,33 @@
   tags:
     - common
 
+- name: preserve hostname
+  sudo: yes
+  lineinfile:
+    dest: /etc/cloud/cloud.cfg
```
Comment (stevendborrelli) [inline roles/common/tasks/main.yml]: Just a question here, why wouldn't we just put in a drop in in `/etc/cloud/config.d`?

**Comment by stevendborrelli** [comment]
Tested on openstack. Looks like it prevents openstack from renaming the system on boot, to LGTM.

File: roles/common/tasks/main.yml (hunk 1 above)
Comment (ryane) [inline roles/common/tasks/main.yml]: just b/c I didn't realize it was an option. that would be better. I'll give it a shot

**Comment by ryane** [comment]
updated to use cloud.cfg.d drop-in file

**Comment by stevendborrelli** [comment]
@kbroughton I am definitely interested in what we can add to terraform.py from ec2.py, especially the multiple tag support and whatever other vars should add.

**Comment by BrianHicks** [comment]
Looks like everything's working with this for me, except some failing distributive stuff. But I think that's outdated, yes?

**Comment by distributorofpain** [comment]
since this is open, i just spun up a mantl setup in DO (rebuilt today) and this time i included a host for kub, but when i go into the webui, i get an internal server error 500 which is related to the merged issue #1367  Is there a way to fix this in a running setup?  i dont fully understand the fix above..

**Pull request merged**